File: erh_core/core/hybrid_model.py
"""
Hybrid Model Module

This module integrates all psychohistory components into a unified framework:
- Temporal ERH
- ABM simulation
- Network dynamics
- Fluid model
- Meta-monitoring

Provides a unified API for running complete psychohistory-style simulations.
"""

import hashlib
import os

import numpy as np
from typing import List, Dict, Optional, Callable, Tuple, Any
from .temporal_erh import track_error_evolution, compute_Pi_temporal, compute_E_temporal
from .abm_simulator import ABMSimulator
from .meta_monitor import MetaMonitor, ERHParameters
from .social_network import SocialNetwork
from .agent import AgentPopulation
from .action_space import Action, generate_world
from .judgement_system import BaseJudge

# Handle imports from analysis module
try:
    from erh_core.analysis.opinion_dynamics import degroot_model, hegselmann_krause_model, aggregate_beliefs
    from erh_core.analysis.fluid_model import solve_error_density_pde, fit_fluid_parameters, detect_critical_phenomena
    from erh_core.analysis.temporal_analysis import analyze_temporal_trends, detect_anomalies, forecast_error_growth
except ImportError:
    # Fallback for test environments or direct execution
    try:
        from ..analysis.opinion_dynamics import degroot_model, hegselmann_krause_model, aggregate_beliefs
        from ..analysis.fluid_model import solve_error_density_pde, fit_fluid_parameters, detect_critical_phenomena
        from ..analysis.temporal_analysis import analyze_temporal_trends, detect_anomalies, forecast_error_growth
    except ImportError:
        from analysis.opinion_dynamics import degroot_model, hegselmann_krause_model, aggregate_beliefs
        from analysis.fluid_model import solve_error_density_pde, fit_fluid_parameters, detect_critical_phenomena
        from analysis.temporal_analysis import analyze_temporal_trends, detect_anomalies, forecast_error_growth


class HybridPsychohistoryModel:
    """
    Integrated psychohistory model combining all components.
    
    This class provides a unified interface for running complete
    psychohistory-style simulations with all features integrated.
    
    Attributes
    ----------
    abm_simulator : ABMSimulator
        ABM simulator component
    meta_monitor : MetaMonitor
        Meta-layer monitor
    temporal_enabled : bool
        Whether temporal tracking is enabled
    network_dynamics_enabled : bool
        Whether network dynamics are enabled
    fluid_model_enabled : bool
        Whether fluid model is enabled
    """
    
    def __init__(
        self,
        num_agents: int = 100,
        judge_factory: Optional[Callable[[int], BaseJudge]] = None,
        network_topology: str = 'small_world',
        enable_temporal: bool = True,
        enable_network_dynamics: bool = True,
        enable_fluid_model: bool = False,
        enable_meta_monitor: bool = True,
        enable_quantum: bool = False,
        quantum_agents_subsample: int = 4,
    ):
        """
        Initialize hybrid model.

        Parameters
        ----------
        num_agents : int, default=100
            Number of agents
        judge_factory : Optional[Callable[[int], BaseJudge]], default=None
            Function to create judges
        network_topology : str, default='small_world'
            Network topology
        enable_temporal : bool, default=True
            Enable temporal ERH tracking
        enable_network_dynamics : bool, default=True
            Enable network opinion dynamics
        enable_fluid_model : bool, default=False
            Enable fluid model (computationally expensive)
        enable_meta_monitor : bool, default=True
            Enable meta-layer monitoring
        enable_quantum : bool, default=False
            Enable quantum VQE stability estimation
        quantum_agents_subsample : int, default=4
            Max agents (qubits) for quantum simulation; subsample if population larger
        """
        # Initialize ABM simulator
        self.abm_simulator = ABMSimulator(
            num_agents=num_agents,
            judge_factory=judge_factory,
            network_topology=network_topology,
            enable_meta_monitor=enable_meta_monitor
        )

        self.meta_monitor = self.abm_simulator.meta_monitor

        # Feature flags
        self.temporal_enabled = enable_temporal
        self.network_dynamics_enabled = enable_network_dynamics
        self.fluid_model_enabled = enable_fluid_model
        self.enable_quantum = enable_quantum
        self.quantum_agents_subsample = quantum_agents_subsample

        # Quantum state (lazy init)
        self._q_sim = None
        self._current_q_params = None
        self._hamiltonian_cache: Dict[str, Any] = {}

        # State
        self.simulation_state = {
            'time': 0,
            'erh_history': [],
            'network_history': [],
            'fluid_history': []
        }

    def _get_interaction_matrix(self, agents: List) -> np.ndarray:
        """Build similarity-based adjacency matrix from agents (error_rate)."""
        n = min(len(agents), self.quantum_agents_subsample)
        if n < 2:
            return np.zeros((1, 1))
        # Subsample uniformly if needed
        if len(agents) > n:
            step = max(1, len(agents) // n)
            idx = list(range(0, len(agents), step))[:n]
            agents = [agents[i] for i in idx]
        else:
            agents = list(agents)[:n]
        matrix = np.zeros((n, n))
        for i in range(n):
            for j in range(n):
                if i < j:
                    w = 1.0 / (1.0 + abs(agents[i].error_rate - agents[j].error_rate))
                    matrix[i, j] = w
                    matrix[j, i] = w
        return matrix

    def _get_biases(self, agents: List) -> np.ndarray:
        """Extract biases from agents (judgment_tendency)."""
        n = min(len(agents), self.quantum_agents_subsample)
        if n < 1:
            return np.array([0.0])
        if len(agents) > n:
            step = max(1, len(agents) // n)
            idx = list(range(0, len(agents), step))[:n]
            agents = [agents[i] for i in idx]
        else:
            agents = list(agents)[:n]
        biases = np.array([np.clip(a.judgment_tendency, -1.0, 1.0) for a in agents])
        return biases

    def _get_cache_key(self, adj_matrix: np.ndarray, agent_data: list) -> str:
        """Generate unique hash for social state (adjacency + agent attributes)."""
        adj_bytes = np.asarray(adj_matrix).tobytes()
        arr = np.array(
            [
                [
                    a.get("empathy", 0.5),
                    a.get("flexibility", 0.5),
                    a.get("resilience", 0.5),
                ]
                for a in agent_data
            ],
            dtype=np.float64,
        )
        return hashlib.md5(adj_bytes + arr.tobytes()).hexdigest()

    def run_simulation(
        self,
        num_time_steps: int = 10,
        actions_per_step: int = 1000,
        tau: float = 0.3,
        X_max: int = 100,
        network_dynamics_model: str = 'degroot',
        fluid_model_params: Optional[Dict] = None
    ) -> Dict:
        """
        Run complete hybrid simulation.
        
        Parameters
        ----------
        num_time_steps : int, default=10
            Number of time steps
        actions_per_step : int, default=1000
            Actions per time step
        tau : float, default=0.3
            Error threshold
        X_max : int, default=100
            Maximum complexity
        network_dynamics_model : str, default='degroot'
            Network dynamics model: 'degroot' or 'hegselmann_krause'
        fluid_model_params : Optional[Dict], default=None
            Parameters for fluid model
            
        Returns
        -------
        Dict
            Complete simulation results
        """
        # Run ABM simulation
        abm_results = self.abm_simulator.run_simulation(
            num_time_steps=num_time_steps,
            actions_per_step=actions_per_step,
            tau=tau,
            X_max=X_max,
            track_erh=self.temporal_enabled
        )
        
        results = {
            'abm_results': abm_results,
            'temporal_erh': None,
            'network_dynamics': None,
            'fluid_model': None,
            'meta_monitoring': None,
            'quantum_stability': None,
        }
        
        # Temporal ERH analysis
        if self.temporal_enabled and 'actions_history' in abm_results:
            temporal_results = self.abm_simulator.compute_temporal_erh(
                abm_results['actions_history'],
                tau=tau,
                X_max=X_max
            )
            results['temporal_erh'] = temporal_results
            
            # Temporal analysis
            if 'E_xt' in temporal_results:
                E_xt = temporal_results['E_xt']
                time_steps = E_xt.shape[0]
                x_values = np.arange(1, X_max + 1)
                
                # Analyze trends
                trends = analyze_temporal_trends(E_xt, time_steps, x_values)
                results['temporal_trends'] = trends
                
                # Detect anomalies
                anomalies = detect_anomalies(E_xt, method='combined', X_max=X_max)
                results['anomalies'] = anomalies
                
                # Forecast
                forecast = forecast_error_growth(E_xt, forecast_horizon=5, X_max=X_max)
                results['forecast'] = forecast
        
        # Network dynamics
        if self.network_dynamics_enabled:
            network = self.abm_simulator.network
            agents = self.abm_simulator.population.agents
            
            if network_dynamics_model == 'degroot':
                dynamics_result = degroot_model(agents, network)
            elif network_dynamics_model == 'hegselmann_krause':
                dynamics_result = hegselmann_krause_model(agents, network)
            else:
                dynamics_result = {'converged': False, 'final_opinions': []}
            
            results['network_dynamics'] = dynamics_result
            
            # Aggregate beliefs
            individual_errors = {agent.agent_id: agent.error_rate for agent in agents}
            aggregated = aggregate_beliefs(individual_errors, network, dynamics_model=network_dynamics_model)
            results['aggregated_beliefs'] = aggregated
        
        # Fluid model (if enabled)
        if self.fluid_model_enabled and self.temporal_enabled and 'E_xt' in results.get('temporal_erh', {}):
            E_xt = results['temporal_erh']['E_xt']
            time_steps = E_xt.shape[0]
            
            # Fit fluid parameters from data
            x_values = np.arange(1, X_max + 1)
            t_values = np.arange(time_steps)
            fluid_params = fit_fluid_parameters(E_xt, x_values, t_values)
            
            # Solve fluid model
            if fluid_model_params is None:
                fluid_model_params = fluid_params
            
            try:
                u_xt, x_grid, t_grid = solve_error_density_pde(
                    x_range=(1, X_max),
                    t_range=(0, time_steps),
                    nx=min(50, X_max),
                    nt=min(50, time_steps),
                    v=fluid_model_params.get('v', 0.1),
                    D=fluid_model_params.get('D', 0.01),
                    alpha=fluid_model_params.get('alpha', 0.05)
                )
                
                # Detect critical phenomena
                critical_events = detect_critical_phenomena(u_xt, x_grid, t_grid)
                
                results['fluid_model'] = {
                    'u_xt': u_xt,
                    'x_grid': x_grid,
                    't_grid': t_grid,
                    'parameters': fluid_params,
                    'critical_events': critical_events
                }
            except Exception as e:
                results['fluid_model'] = {'error': str(e)}
        
        # Meta-monitoring summary
        if self.meta_monitor:
            results['meta_monitoring'] = self.meta_monitor.get_monitoring_summary()

        # Quantum Ethical Hilbert Space (AdvancedEthicalQuantumEngine)
        if self.enable_quantum:
            try:
                from simulation.quantum.simulator import AdvancedEthicalQuantumEngine

                agents = self.abm_simulator.population.agents
                if agents:
                    n_q = min(len(agents), self.quantum_agents_subsample, 20)
                    n_q = max(n_q, 2)
                    if self._q_sim is None:
                        use_ibm = os.environ.get("USE_IBM_QUANTUM", "").lower() in ("1", "true", "yes")
                        backend_name = os.environ.get("IBM_QUANTUM_BACKEND", "ibm_fez")
                        self._q_sim = AdvancedEthicalQuantumEngine(
                            num_agents=n_q,
                            use_real_hardware=use_ibm,
                            backend_name=backend_name,
                        )

                    active_agents = agents[:n_q] if len(agents) <= n_q else list(agents[i] for i in range(0, len(agents), max(1, len(agents) // n_q)))[:n_q]
                    agent_data = [
                        {
                            "empathy": getattr(a, "empathy", 1.0 - a.error_rate),
                            "flexibility": getattr(a, "flexibility", 0.5 + a.judgment_tendency / 2),
                            "resilience": getattr(a, "resilience", 1.0 - a.error_rate),
                        }
                        for a in active_agents
                    ]

                    network = self.abm_simulator.network
                    if hasattr(network, "get_adjacency_submatrix"):
                        adj_matrix = network.get_adjacency_submatrix(n_q)
                    else:
                        adj_matrix = self._get_interaction_matrix(active_agents)

                    cache_key = self._get_cache_key(adj_matrix, agent_data)
                    if cache_key in self._hamiltonian_cache:
                        q_results = self._hamiltonian_cache[cache_key]
                    else:
                        root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
                        fig_dir = os.path.join(root, "simulation", "output", "figures")
                        q_results = self._q_sim.run_simulation(
                            agent_data, adj_matrix, output_dir=fig_dir
                        )
                        self._hamiltonian_cache[cache_key] = q_results

                    # Add quantum_energy (social_tension) via SocialDynamicsQuantumSimulator
                    biases = np.array([a.get("flexibility", 0.5) - 0.5 for a in agent_data], dtype=float)
                    try:
                        from simulation.quantum.simulator import SocialDynamicsQuantumSimulator

                        q_sim_ising = SocialDynamicsQuantumSimulator(
                            num_agents=adj_matrix.shape[0] if adj_matrix.size else 2,
                            topology="full",
                            seed=42,
                        )
                        quantum_energy = q_sim_ising.measure_social_tension(adj_matrix, biases)
                        q_results["quantum_energy"] = quantum_energy
                        q_results["social_tension_energy"] = quantum_energy
                        q_results["system_energy"] = quantum_energy
                    except Exception:
                        pass
                    q_results["magnetization"] = q_results.get("magnetization", q_results.get("system_coherence", 0.0))

                    if "raw_counts" in q_results:
                        try:
                            from erh_core.analysis.statistics import calculate_von_neumann_entropy

                            counts = q_results["raw_counts"]
                            if not counts:
                                raise ValueError("empty counts")
                            total = sum(counts.values())
                            n_qubits = len(next(iter(counts.keys())))
                            dim = 2 ** n_qubits
                            rho = np.zeros((dim, dim), dtype=np.float64)
                            for outcome, c in counts.items():
                                idx = int(outcome, 2) if outcome else 0
                                rho[idx, idx] = c / total if total else 0.0
                            q_results["von_neumann_entropy"] = calculate_von_neumann_entropy(rho)
                        except Exception:
                            pass
                    results["quantum_stability"] = q_results
            except ImportError:
                results["quantum_stability"] = {"error": "simulation.quantum not available"}
            except Exception as e:
                results["quantum_stability"] = {"error": str(e)}

        # Update state and simulation_history (C2.2)
        self.simulation_state['time'] = num_time_steps
        self.simulation_state['erh_history'] = abm_results.get('erh_history', [])
        if results.get('quantum_stability') and isinstance(results['quantum_stability'], dict):
            qs = results['quantum_stability']
            qe = qs.get('quantum_energy', qs.get('social_tension_energy'))
            vne = qs.get('von_neumann_entropy')
            self.simulation_state['quantum_energy'] = qe
            self.simulation_state['von_neumann_entropy'] = vne
            # Write to last simulation_history entry (C2.2)
            if self.abm_simulator.simulation_history:
                last = self.abm_simulator.simulation_history[-1]
                if isinstance(last, dict):
                    last['quantum_energy'] = qe
                    last['system_energy'] = qe
                    last['von_neumann_entropy'] = vne
                    last['magnetization'] = qs.get('magnetization', qs.get('system_coherence', 0.0))

        return results
    
    def adaptive_adjustment(
        self,
        simulation_results: Dict,
        target_exponent: float = 0.5
    ) -> Dict:
        """
        Perform adaptive adjustment based on simulation results.
        
        Parameters
        ----------
        simulation_results : Dict
            Results from run_simulation
        target_exponent : float, default=0.5
            Target ERH exponent
            
        Returns
        -------
        Dict
            Adjustment results and recommendations
        """
        adjustments = {}
        
        # Meta-monitor adaptive parameters
        if self.meta_monitor and 'temporal_erh' in simulation_results:
            temporal_erh = simulation_results['temporal_erh']
            if 'E_xt' in temporal_erh:
                E_xt = temporal_erh['E_xt']
                # Use meta-monitor's adaptive adjustment
                E_xt_history = [E_xt]
                new_params = self.meta_monitor.adaptive_erh_parameters(E_xt_history, target_exponent)
                adjustments['erh_parameters'] = {
                    'C': new_params.C,
                    'epsilon': new_params.epsilon
                }
        
        # ABM calibration
        if 'abm_results' in simulation_results:
            calibration = self.abm_simulator.calibrate_erh_parameters(
                simulation_results['abm_results'],
                target_exponent
            )
            adjustments['abm_calibration'] = calibration
        
        return adjustments
    
    def get_unified_metrics(self, simulation_results: Dict) -> Dict:
        """
        Compute unified metrics across all components.
        
        Parameters
        ----------
        simulation_results : Dict
            Simulation results
            
        Returns
        -------
        Dict
            Unified metrics
        """
        metrics = {
            'erh_satisfaction': None,
            'temporal_stability': None,
            'network_coherence': None,
            'system_health': None
        }
        
        # ERH satisfaction
        if 'temporal_erh' in simulation_results:
            temporal_erh = simulation_results['temporal_erh']
            if 'erh_satisfaction' in temporal_erh:
                erh_sat = temporal_erh['erh_satisfaction']
                metrics['erh_satisfaction'] = {
                    'satisfaction_rate': erh_sat.get('satisfaction_rate', 0.0),
                    'violation_rate': erh_sat.get('violation_rate', 0.0),
                    'worst_violation': erh_sat.get('worst_violation', 0.0)
                }
        
        # Temporal stability
        if 'temporal_trends' in simulation_results:
            trends = simulation_results['temporal_trends']
            overall = trends.get('overall_trend', {})
            metrics['temporal_stability'] = {
                'volatility': overall.get('volatility', 0.0),
                'trend_direction': overall.get('direction', 'unknown'),
                'mean_error': overall.get('mean_error', 0.0)
            }
        
        # Network coherence
        if 'network_dynamics' in simulation_results:
            network_dyn = simulation_results['network_dynamics']
            metrics['network_coherence'] = {
                'converged': network_dyn.get('converged', False),
                'iterations': network_dyn.get('iterations', 0),
                'clusters': len(network_dyn.get('clusters', []))
            }
        
        # System health (composite metric)
        health_score = 1.0
        
        if metrics['erh_satisfaction']:
            satisfaction_rate = metrics['erh_satisfaction']['satisfaction_rate']
            health_score *= satisfaction_rate
        
        if metrics['temporal_stability']:
            volatility = metrics['temporal_stability']['volatility']
            # Lower volatility is better
            health_score *= max(0.0, 1.0 - volatility / 0.5)
        
        if metrics['network_coherence']:
            if metrics['network_coherence']['converged']:
                health_score *= 1.1  # Bonus for convergence
            else:
                health_score *= 0.9
        
        metrics['system_health'] = {
            'score': min(1.0, max(0.0, health_score)),
            'status': 'healthy' if health_score > 0.7 else 'degraded' if health_score > 0.4 else 'critical'
        }
        
        return metrics
    
    def get_summary(self) -> Dict:
        """
        Get summary of hybrid model state.
        
        Returns
        -------
        Dict
            Model summary
        """
        return {
            'num_agents': len(self.abm_simulator.population),
            'network_topology': self.abm_simulator.network.get_network_statistics(),
            'features_enabled': {
                'temporal': self.temporal_enabled,
                'network_dynamics': self.network_dynamics_enabled,
                'fluid_model': self.fluid_model_enabled,
                'meta_monitor': self.meta_monitor is not None,
                'quantum': self.enable_quantum,
            },
            'simulation_state': self.simulation_state,
            'abm_summary': self.abm_simulator.get_simulation_summary()
        }

File: julia/src/HybridModel.jl
"""
    HybridModel — Integrated psychohistory model combining all simulation components.

    Mirrors erh_core/core/hybrid_model.py.

    Combines:
    - ABMSimulator      (agent-based simulation)
    - SocialNetwork     (graph structure + opinion dynamics)
    - TemporalERH       (temporal Π/B/E tracking)
    - FluidModel        (PDE-based error density evolution, optional)

    Note: the quantum component (AdvancedEthicalQuantumEngine) is handled by
    QuantumSimulator.jl and is exposed via an optional `enable_quantum` flag
    that is a no-op until Phase 3 files are compiled.
"""
module HybridModel

using Statistics
using LinearAlgebra

# Sibling modules (resolved at ERH module level via include order in ERH.jl)
import ..ABMSimulator as ABMS
import ..SocialNetwork as SN
import ..TemporalERH as TERH
import ..FluidModel as FM

export HybridPsychohistoryModel, create_hybrid_model,
       run_hybrid_simulation!, adaptive_adjustment,
       get_unified_metrics, get_hybrid_summary

# ---------------------------------------------------------------------------
# Config struct
# ---------------------------------------------------------------------------

"""
    HybridConfig

Feature flags and hyperparameters for the hybrid model.
"""
struct HybridConfig
    num_agents::Int
    network_topology::String
    enable_temporal::Bool
    enable_network_dynamics::Bool
    enable_fluid_model::Bool
    enable_meta_monitor::Bool
    enable_quantum::Bool
    network_dynamics_model::String   # "degroot" | "hegselmann_krause"
    seed::Int
end

function HybridConfig(;
    num_agents::Int=100,
    network_topology::String="small_world",
    enable_temporal::Bool=true,
    enable_network_dynamics::Bool=true,
    enable_fluid_model::Bool=false,
    enable_meta_monitor::Bool=true,
    enable_quantum::Bool=false,
    network_dynamics_model::String="degroot",
    seed::Int=42
)
    HybridConfig(num_agents, network_topology, enable_temporal,
                 enable_network_dynamics, enable_fluid_model,
                 enable_meta_monitor, enable_quantum,
                 network_dynamics_model, seed)
end

# ---------------------------------------------------------------------------
# Main model struct
# ---------------------------------------------------------------------------

"""
    HybridPsychohistoryModel

Holds the ABM simulator, the social network (lazily initialised from ABM state),
and runtime configuration.
"""
mutable struct HybridPsychohistoryModel
    config::HybridConfig
    abm::ABMS.ABMSimulatorState
    network::Union{SN.SocialNetworkState, Nothing}
    simulation_state::Dict{String, Any}
end

# ---------------------------------------------------------------------------
# Constructor
# ---------------------------------------------------------------------------

"""
    create_hybrid_model(; kwargs...) → HybridPsychohistoryModel

Convenience constructor.  All kwargs are forwarded to `HybridConfig`.
"""
function create_hybrid_model(; kwargs...)
    cfg = HybridConfig(; kwargs...)
    abm = ABMS.create_simulator(;
        num_agents       = cfg.num_agents,
        network_topology = cfg.network_topology,
        seed             = cfg.seed
    )

    # Build the social network from initial ABM agent state
    using Agents: allagents
    agents_vec = collect(allagents(abm.model))
    agent_ids  = [a.id for a in agents_vec]
    error_rates = [a.error_rate for a in agents_vec]
    tendencies  = [a.judgment_tendency for a in agents_vec]

    network = SN.create_network(
        agent_ids, error_rates, tendencies;
        topology = cfg.network_topology,
        seed     = cfg.seed
    )

    sim_state = Dict{String,Any}(
        "time"          => 0,
        "erh_history"   => Any[],
        "network_history" => Any[],
        "fluid_history" => Any[]
    )

    HybridPsychohistoryModel(cfg, abm, network, sim_state)
end

# ---------------------------------------------------------------------------
# Main simulation entry point
# ---------------------------------------------------------------------------

"""
    run_hybrid_simulation!(model; num_time_steps, actions_per_step, τ, X_max,
                            fluid_model_params) → Dict{String,Any}

Run a complete hybrid simulation and return a comprehensive results dict.

Keys in the returned dict:
- `"abm_results"`        — raw output from ABMSimulator
- `"temporal_erh"`       — Π/B/E arrays (if `enable_temporal`)
- `"network_dynamics"`   — opinion dynamics result (if `enable_network_dynamics`)
- `"aggregated_beliefs"` — per-agent aggregated errors
- `"fluid_model"`        — PDE solution dict (if `enable_fluid_model`)
"""
function run_hybrid_simulation!(
    model::HybridPsychohistoryModel;
    num_time_steps::Int=10,
    actions_per_step::Int=1000,
    τ::Float64=0.3,
    X_max::Int=100,
    fluid_model_params::Union{Dict{String,Float64}, Nothing}=nothing
)::Dict{String, Any}

    # --- ABM ---
    abm_results = ABMS.run_simulation!(
        model.abm;
        num_time_steps       = num_time_steps,
        actions_per_step     = actions_per_step,
        τ                    = τ,
        X_max                = X_max,
        track_erh            = model.config.enable_temporal,
        interaction_probability = 0.1,
        seed                 = model.config.seed
    )

    results = Dict{String,Any}(
        "abm_results"        => abm_results,
        "temporal_erh"       => nothing,
        "network_dynamics"   => nothing,
        "aggregated_beliefs" => nothing,
        "fluid_model"        => nothing,
        "quantum_stability"  => nothing
    )

    # --- Temporal ERH ---
    if model.config.enable_temporal && haskey(abm_results, "actions_history")
        temporal = ABMS.compute_temporal_erh(
            model.abm,
            abm_results["actions_history"];
            τ    = τ,
            X_max = X_max
        )
        results["temporal_erh"] = temporal
    end

    # --- Network dynamics ---
    if model.config.enable_network_dynamics && model.network !== nothing
        using Agents: allagents
        agents_vec = collect(allagents(model.abm.model))
        tendencies  = [a.judgment_tendency for a in agents_vec]
        agent_ids   = [a.id for a in agents_vec]
        error_rates = [a.error_rate for a in agents_vec]

        # Sync network node attributes
        SN.update_node_attributes!(model.network, agent_ids, error_rates, tendencies)

        dyn_result = if model.config.network_dynamics_model == "degroot"
            SN.degroot_model!(tendencies, model.network)
        else
            SN.hegselmann_krause_model!(tendencies, model.network)
        end
        results["network_dynamics"] = dyn_result

        individual_errors = Dict(agent_ids[i] => error_rates[i] for i in eachindex(agent_ids))
        aggregated = SN.aggregate_beliefs!(
            individual_errors, model.network;
            dynamics_model = model.config.network_dynamics_model
        )
        results["aggregated_beliefs"] = aggregated
    end

    # --- Fluid model ---
    if model.config.enable_fluid_model
        temporal = results["temporal_erh"]
        if temporal !== nothing && haskey(temporal, "E_xt")
            E_xt = temporal["E_xt"]
            T    = size(E_xt, 1)
            x_vals = collect(1.0:Float64(X_max))
            t_vals = collect(0.0:Float64(T-1))

            fitted = FM.fit_fluid_parameters(E_xt, x_vals, t_vals)
            params = fluid_model_params !== nothing ? fluid_model_params : fitted

            try
                nx_fluid = min(50, X_max)
                nt_fluid = min(50, T)
                u_xt, x_grid, t_grid = FM.solve_error_density_pde(
                    (1.0, Float64(X_max)), (0.0, Float64(T));
                    nx = nx_fluid, nt = nt_fluid,
                    v  = get(params, "v",     0.1),
                    D  = get(params, "D",     0.01),
                    alpha = get(params, "alpha", 0.05)
                )
                critical = FM.detect_critical_phenomena(u_xt, x_grid, t_grid)
                results["fluid_model"] = Dict{String,Any}(
                    "u_xt"           => u_xt,
                    "x_grid"         => x_grid,
                    "t_grid"         => t_grid,
                    "parameters"     => params,
                    "critical_events"=> critical
                )
            catch e
                results["fluid_model"] = Dict{String,Any}("error" => string(e))
            end
        end
    end

    # Update state
    model.simulation_state["time"]        = num_time_steps
    model.simulation_state["erh_history"] = get(abm_results, "erh_history", Any[])

    return results
end

# ---------------------------------------------------------------------------
# Adaptive adjustment
# ---------------------------------------------------------------------------

"""
    adaptive_adjustment(model, simulation_results; target_exponent) → Dict{String,Any}

Calibrate ERH parameters and surface recommendations based on simulation output.
"""
function adaptive_adjustment(
    model::HybridPsychohistoryModel,
    simulation_results::Dict{String,Any};
    target_exponent::Float64=0.5
)::Dict{String,Any}
    adjustments = Dict{String,Any}()

    # ABM-level calibration
    if haskey(simulation_results, "abm_results")
        calibration = ABMS.calibrate_erh_parameters(
            simulation_results["abm_results"];
            target_exponent = target_exponent
        )
        adjustments["abm_calibration"] = calibration
    end

    # Temporal ERH parameter scan
    if haskey(simulation_results, "temporal_erh") &&
       simulation_results["temporal_erh"] !== nothing
        terh = simulation_results["temporal_erh"]
        if haskey(terh, "E_xt")
            E_xt = terh["E_xt"]
            T, Xn = size(E_xt)
            # Estimate global exponent from |E_xt| vs x
            x_vals  = Float64.(1:Xn)
            abs_E   = abs.(E_xt)
            mean_E  = vec(mean(abs_E, dims=1))
            pos     = findall(e -> e > 0, mean_E)
            if length(pos) >= 3
                lx = log.(x_vals[pos])
                lE = log.(mean_E[pos])
                mx, mE2 = mean(lx), mean(lE)
                gamma = sum((lx .- mx) .* (lE .- mE2)) /
                        (sum((lx .- mx).^2) + 1e-10)
                adjustments["estimated_global_exponent"] = gamma
                adjustments["erh_parameters"] = Dict(
                    "C"       => 1.0,
                    "epsilon" => max(0.0, gamma - 0.5)
                )
            end
        end
    end

    return adjustments
end

# ---------------------------------------------------------------------------
# Unified metrics
# ---------------------------------------------------------------------------

"""
    get_unified_metrics(simulation_results) → Dict{String,Any}

Compute composite metrics across ABM, temporal ERH, and network dynamics.
"""
function get_unified_metrics(simulation_results::Dict{String,Any})::Dict{String,Any}
    metrics = Dict{String,Any}(
        "erh_satisfaction"   => nothing,
        "temporal_stability" => nothing,
        "network_coherence"  => nothing,
        "system_health"      => nothing
    )

    # ERH satisfaction
    terh = get(simulation_results, "temporal_erh", nothing)
    if terh !== nothing && haskey(terh, "erh_satisfaction")
        sat = terh["erh_satisfaction"]
        metrics["erh_satisfaction"] = Dict(
            "satisfaction_rate" => get(sat, "satisfaction_rate", 0.0),
            "violation_rate"    => get(sat, "violation_rate",    0.0)
        )
    end

    # Temporal stability (volatility of E_xt)
    if terh !== nothing && haskey(terh, "E_xt")
        E_xt = terh["E_xt"]
        abs_E = abs.(E_xt)
        volatility = std(abs_E)
        metrics["temporal_stability"] = Dict(
            "volatility"  => volatility,
            "mean_error"  => mean(abs_E)
        )
    end

    # Network coherence
    net_dyn = get(simulation_results, "network_dynamics", nothing)
    if net_dyn !== nothing
        metrics["network_coherence"] = Dict(
            "converged"  => net_dyn.converged,
            "iterations" => net_dyn.iterations,
            "clusters"   => hasfield(typeof(net_dyn), :clusters) ?
                            length(net_dyn.clusters) : 0
        )
    end

    # System health composite
    health_score = 1.0
    if metrics["erh_satisfaction"] !== nothing
        health_score *= get(metrics["erh_satisfaction"], "satisfaction_rate", 1.0)
    end
    if metrics["temporal_stability"] !== nothing
        vol = get(metrics["temporal_stability"], "volatility", 0.0)
        health_score *= max(0.0, 1.0 - vol / 0.5)
    end
    if metrics["network_coherence"] !== nothing
        health_score *= get(metrics["network_coherence"], "converged", false) ? 1.1 : 0.9
    end

    score  = clamp(health_score, 0.0, 1.0)
    status = score > 0.7 ? "healthy" : score > 0.4 ? "degraded" : "critical"
    metrics["system_health"] = Dict("score" => score, "status" => status)

    return metrics
end

# ---------------------------------------------------------------------------
# Summary
# ---------------------------------------------------------------------------

"""
    get_hybrid_summary(model) → Dict{String,Any}
"""
function get_hybrid_summary(model::HybridPsychohistoryModel)::Dict{String,Any}
    abm_summary = ABMS.get_simulation_summary(model.abm)
    net_stats   = model.network !== nothing ?
                  SN.get_network_statistics(model.network) :
                  Dict{String,Any}()

    return Dict{String,Any}(
        "num_agents"       => model.config.num_agents,
        "network_topology" => model.config.network_topology,
        "features_enabled" => Dict(
            "temporal"         => model.config.enable_temporal,
            "network_dynamics" => model.config.enable_network_dynamics,
            "fluid_model"      => model.config.enable_fluid_model,
            "meta_monitor"     => model.config.enable_meta_monitor,
            "quantum"          => model.config.enable_quantum
        ),
        "simulation_state" => model.simulation_state,
        "abm_summary"      => abm_summary,
        "network_stats"    => net_stats
    )
end

end # module HybridModel
